### PR TITLE
mrc-1654: better handling of integer vectors in sum and rmhyper

### PR DESCRIPTION
**Pull request by richfitz**

### CODE REVIEW

File: tests/testthat/run/test-run-general.R [hunk 1]
```diff
@@ -1245,3 +1245,25 @@ test_that("user parameter validation", {
...
     "Unknown user parameters: x")
 })
+
+test_that("sum over integer", {
+  gen <- odin({
+    x[] <- user()
+    dim(x) <- user()
+    idx[] <- user()
+    dim(idx) <- user()
+    initial(v[]) <- x[idx[i]] # TODO: fixme
```
Comment (r-ash) [inline tests/testthat/run/test-run-general.R:1255]: Old TODO?

File: R/generate_c_equation.R [hunk 2]
```diff
@@ -203,7 +203,7 @@ generate_c_equation_copy <- function(eq, data_info, dat, rewrite) {
                    target, lhs, len, data_info$storage_type)
     } else {
       offset <- rewrite(x$offset)
-      c(sprintf_safe("for (size_t i = 0; i < %s; ++i) {", len),
+      c(sprintf_safe("for (int i = 0; i < %s; ++i) {", len),
```
Comment (r-ash) [inline R/generate_c_equation.R:206]: Why did you chose to switch to `int` here where other for loops introduced in this PR are using `size_t`? Probably not a big deal, or maybe some detail of C I don't understand.

**Comment by r-ash** [comment]
> This looks fine, one TODO sitting around which probably needs removing but looks good besides that.
> 
> Also interested in what required this change though? Does this fix some bug?

Oh looking at the actual ticket I guess something was generating code calling `odin_sum1` which was trying to pass an int in?

File: tests/testthat/run/test-run-general.R (hunk 1 above)
Comment (richfitz) [reply tests/testthat/run/test-run-general.R:1255]: I have absolutely no idea what this TODO is about

Comment (richfitz) [reply tests/testthat/run/test-run-general.R:1255]: Oh looks like it was copied from an example above

**Comment by richfitz** [comment]
> Also interested in what required this change though? Does this fix some bug?

The code in question all had the form

```
f(double* x, ...)
f(data)
```

which works fine when `data` is an array of doubles, but really badly when `data` is an array of ints. The C compiler passes the data, with a warning - that casts the data to a type that they are not and you get complete nonsense out, fairly silently.

If you run the tests against an old odin version they fail, and if you put `verbose = TRUE` in you'll see the warning

**Pull request merged**